bundle/src/main/java/com/adobe/acs/commons/remoteassets/impl/RemoteAssetDecorator.java

```diff
@@ -0,0 +1,175 @@
+
+    /**
+     * This set stores resource paths for remote assets that are in the process
+     * of being sync'd from the remote server.  This prevents an infinite loop
+     * when the RemoteAssetSync service fetches the asset in order to update it.
+     */
+    private static Set<String> remoteResourcesSyncing = new ConcurrentSkipListSet<>();
+
+    @Reference
+    private RemoteAssetsBinarySync assetSync;
+
+    @Reference
+    private RemoteAssetsConfig config;
+
+    /**
+     * When resolving a remote asset, first sync the asset from the remote server.
+     * @param resource The resource being resolved.
+     * @return The current resource.  If the resource is a "remote" asset, it will
+     * first be converted to a true local AEM asset by sync'ing in the rendition
+     * binaries from the remote server.
+     */
+    @Override
+    public Resource decorate(final Resource resource) {
+        try {
+            if (!this.accepts(resource)) {
+                return resource;
+            }
+        } catch (Exception e) {
+            // Logging at debug level b/c if this happens it could represent a ton of logging
+            if (LOG.isDebugEnabled()) {
+                LOG.debug("Failed binary sync check for remote asset: {} - {}", resource.getPath(), e.getMessage());
+            }
+            return resource;
+        }
+
+        Resource ret = resource;
+        try {
+            remoteResourcesSyncing.add(resource.getPath());
+            LOG.info("Sync'ing remote asset binaries: {}", resource.getPath());
+            ret = this.assetSync.syncAsset(resource);
+        } catch (Exception e) {
+            LOG.error("Failed to sync binaries for remote asset: {} - {}", resource.getPath(), e.getMessage());
+        } finally {
+            remoteResourcesSyncing.remove(resource.getPath());
+        }
+        return ret;
+    }
+
+    /**
+     * @deprecated
+     * When resolving a remote asset, first sync the asset from the remote server.
+     * @param resource The resource being resolved.
+     * @param request HttpServletRequest
+     * @return The current resource.  If the resource is a "remote" asset, it will
+     * first be converted to a true local AEM asset by sync'ing in the rendition
+     * binaries from the remote server.
+     */
+    @Deprecated
+    @Override
+    public Resource decorate(final Resource resource, final HttpServletRequest request) {
+        return this.decorate(resource);
+    }
+
+    /**
+     * Check if this resource is a remote resource.
+     * @param resource Resource to check
+     * @return true if resource is remote, else false
+     */
+    private boolean accepts(final Resource resource) throws RepositoryException {
+        if (resource == null) {
+            return false;
+        }
+
+        ValueMap props = resource.getValueMap();
+        if (!props.get("jcr:primaryType", "").equals("dam:AssetContent")) {
```

Comment:
Use constants? JcrConstants and DamConstants? .. sort of feels weird w the order, since dam:AssetContent is as static string and thus impossible to ever return null.. where as .get() isa method and if the params changed it could.

Reply:
Are you asking for this to be `if (!DamCostants.NT_DAM_ASSETCONTENT.equals(props.get(JcrConstants.JCR_PRIMARYTYPE)) {`?